interpreter/prune.go

```diff
@@ -162,79 +161,124 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
 	switch node.ExprKind.(type) {
 	case *exprpb.Expr_SelectExpr:
 		if operand, pruned := p.prune(node.GetSelectExpr().Operand); pruned {
-			newExpr := *node
-			newSelect := *newExpr.GetSelectExpr()
-			newSelect.Operand = operand
-			newExpr.GetExprKind().(*exprpb.Expr_SelectExpr).SelectExpr = &newSelect
-			return &newExpr, true
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_SelectExpr{
+					SelectExpr: &exprpb.Expr_Select{
+						Operand:  operand,
+						Field:    node.GetSelectExpr().GetField(),
+						TestOnly: node.GetSelectExpr().GetTestOnly(),
+					},
+				},
+			}, true
 		}
 	case *exprpb.Expr_CallExpr:
 		if newExpr, pruned := p.maybePruneFunction(node); pruned {
 			newExpr, _ = p.prune(newExpr)
 			return newExpr, true
 		}
-		newCall := *node.GetCallExpr()
 		var prunedCall bool
-		var prunedArg bool
-		for i, arg := range node.GetCallExpr().Args {
-			if newCall.Args[i], prunedArg = p.prune(arg); prunedArg {
+		call := node.GetCallExpr()
+		args := call.GetArgs()
+		newArgs := make([]*exprpb.Expr, len(args))
+		newCall := &exprpb.Expr_Call{
+			Function: call.GetFunction(),
+			Target:   call.GetTarget(),
+			Args:     newArgs,
+		}
+		for i, arg := range args {
+			newArgs[i] = args[i]
+			if newArg, prunedArg := p.prune(arg); prunedArg {
 				prunedCall = true
+				newArgs[i] = newArg
 			}
 		}
-		if newTarget, prunedTarget := p.prune(node.GetCallExpr().Target); prunedTarget {
+		if newTarget, prunedTarget := p.prune(call.GetTarget()); prunedTarget {
 			prunedCall = true
 			newCall.Target = newTarget
 		}
 		if prunedCall {
-			newExpr := *node
-			newExpr.GetExprKind().(*exprpb.Expr_CallExpr).CallExpr = &newCall
-			return &newExpr, true
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_CallExpr{
+					CallExpr: newCall,
+				},
+			}, true
 		}
 	case *exprpb.Expr_ListExpr:
-		newList := *node.GetListExpr()
+		elems := node.GetListExpr().GetElements()
+		newElems := make([]*exprpb.Expr, len(elems))
 		var prunedList bool
-		var prunedElem bool
-		for i, elem := range node.GetListExpr().Elements {
-			if newList.Elements[i], prunedElem = p.prune(elem); prunedElem {
+		for i, elem := range elems {
+			if newElem, prunedElem := p.prune(elem); prunedElem {
```

Comment:
It looks like you also need to copy the non-pruned elements too, just in case a later element in the list is pruned. Otherwise your list will have a bunch of nulls. Please add a test for this case.